<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title>‚öñÔ∏è How class imbalance might affect genetic demultiplexing of scRNA-seq data | Hoang (Anh) Tran</title> <meta name="author" content="Hoang (Anh) Tran"> <meta name="description" content="Benchmarking souporcell for scRNA-seq data in the case of class imbalance"> <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.css"> <link rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?19f3075a2d19613090fe9e16b564e1fe" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%E2%9A%9B%EF%B8%8F&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://hoangmgh.github.io/blog/2023/benchmarking-souporcell/"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?e74e74bf055e5729d44a7d031a5ca6a5" media="none" id="highlight_theme_dark"> <script src="/assets/js/theme.js?96d6b3e1c3604aca8b6134c7afdd5db6"></script> <script src="/assets/js/dark_mode.js?9b17307bb950ffa2e34be0227f53558f"></script> <script async src="https://www.googletagmanager.com/gtag/js?id={{%20site.google_analytics%20}}"></script> <script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-1W7V2Y0MXW");</script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"><span class="font-weight-bold">Hoang¬†</span>(Anh)¬†Tran</a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about</a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">Bioinformatics<span class="sr-only">(current)</span></a> </li> <li class="nav-item "> <a class="nav-link" href="/projects/">Arts</a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications</a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv</a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="fa-solid fa-moon"></i> <i class="fa-solid fa-sun"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5"> <div class="post"> <header class="post-header"> <h1 class="post-title">‚öñÔ∏è How class imbalance might affect genetic demultiplexing of scRNA-seq data</h1> <p class="post-meta">October 31, 2023</p> <p class="post-tags"> <a href="/blog/2023"> <i class="fa-solid fa-calendar fa-sm"></i> 2023 </a> ¬† ¬∑ ¬† <a href="/blog/tag/genetic-demultiplexing"> <i class="fa-solid fa-hashtag fa-sm"></i> genetic-demultiplexing</a> ¬† <a href="/blog/tag/troubleshooting"> <i class="fa-solid fa-hashtag fa-sm"></i> troubleshooting</a> ¬† ¬† ¬∑ ¬† <a href="/blog/category/sample-posts"> <i class="fa-solid fa-tag fa-sm"></i> sample-posts</a> ¬† </p> </header> <article class="post-content"> <div id="markdown-content"> <font size="5">1. Troubleshooting previous known problems </font> <p>From my experience with souporcell, I have seen that in some cases, one donor might be completely missing from the assignments. Instead of having 8 assigned donors, for example, you will have 7 donors with an equal number of cells, while one donor have around 20 cells - almost nothing. For example, here is one of our earlier experiment, where each cell has a hashtag identifying its origin (HTO), and we also used souporcell to assign a cell to a donor:</p> <div class="row" style="text-align: center"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" media="(max-width: 480px)" srcset="/assets/img/soup_misassignment_example-480.webp"></source> <source class="responsive-img-srcset" media="(max-width: 800px)" srcset="/assets/img/soup_misassignment_example-800.webp"></source> <source class="responsive-img-srcset" media="(max-width: 1400px)" srcset="/assets/img/soup_misassignment_example-1400.webp"></source> <img src="/assets/img/soup_misassignment_example.png" class="img-fluid rounded z-depth-1" width="auto" height="auto" title="Soup missasignment example" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Fig.1: An example illustrating missing donors from souporcell experiments </div> <p>In two of our batches, we have 8 samples: the columns are the identities given by the hashtag, and the rows are the assignments from souporcell. As you can see, $AO39PO$ and $AO38PO$ are merged as Donor2, while Donor7 is trivial. The same phenomenon is also evident in the other Batch, DIA_3P_3C. What is similar in both cases is the low number of cells from the messed up Donors, i.e. 200-500 cells. By looking at the formula of the loss function and the fact that gaussian mixture might be very prone to class imbalance, I suspect that this is really the case.</p> <p><strong>2. Baseline scenario</strong> And so, I attempted to simulate this situation by using our in-house data. From our Thyroid Autoimmune disease scRNA-seq dataset, I extracted the top 2000 cells from 7 Donors. All of these are CD45+ cells extracted from the Thyroid tissue. These cells are the one with the highest UMI and $n_genes$ expressed, so at least we can be ascertain that coverage is not the problem. As a baseline, I run souporcell demultiplexing with 14,000 cells. In one run, here is the result that I have:</p> <div class="row" style="text-align: center"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" media="(max-width: 480px)" srcset="/assets/img/14k_cells_soup-480.webp"></source> <source class="responsive-img-srcset" media="(max-width: 800px)" srcset="/assets/img/14k_cells_soup-800.webp"></source> <source class="responsive-img-srcset" media="(max-width: 1400px)" srcset="/assets/img/14k_cells_soup-1400.webp"></source> <img src="/assets/img/14k_cells_soup.png" class="img-fluid rounded z-depth-1" width="auto" height="auto" title="Soup missasignment example" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Fig.2: An example showing the stochastic nature of the algorithm </div> <p>This is quite interesting! I expect that most of the cells will fall into its desired bin, i.e. the diagonal entries will consist each of approximately 2000 cells. Instead, what I see is the merging of 2 donors into one assignment, and the splitting of one authentic donor into two clusters! ‚ùó‚ùó<strong>This is very alarming!</strong> given that most people will take the souporcell result at face‚Äôs value. However, the truth is that you are seeing just 6 out of 7 donors.</p> <p>I then went on to repeat the same run again, but this time with a different parameter. Instead of using $1024$ loci as the maximum number of features, I used $200$ loci. Hopefully this can fix the issue! Indeed, we observe in the next run that most of the cells got accurately assigned:</p> <div class="row"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" media="(max-width: 480px)" srcset="/assets/img/soup_good_example-480.webp"></source> <source class="responsive-img-srcset" media="(max-width: 800px)" srcset="/assets/img/soup_good_example-800.webp"></source> <source class="responsive-img-srcset" media="(max-width: 1400px)" srcset="/assets/img/soup_good_example-1400.webp"></source> <img src="/assets/img/soup_good_example.png" class="img-fluid rounded z-depth-1" width="auto" height="auto" title="Soup missasignment example" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Fig.3: A highly accurate result from one run with 14k cells, 2000 from each of 7 donors </div> <p>I think this is quite alarming, because most labs will likely run souporcell one time for one experiment and accept the result as-is. I also suspect that if you run this multiple times, perhaps you will arrive at a better solution than the one given in Fig.2 However, it might be true that one single run might give a suboptimal solution, local minima. When one tune the parameter drastically, the result changed, even though it might look like we have somewhat an equal number of cells.</p> <p><strong>Imbalance scenario 1:</strong> Having finished this experiment, I went on to simulate an imbalance situation, where I subset 500 cells randomly from one donor, and keep 2000 cells from each of 7 donors. I then ran souporcell clustering procedure with the following parameters: \(n_{loci} = 200\), \(n_{loci} = 1024\), each with several repeats (\(n_{repeats} = 50\)):</p> <div class="row" style="text-align: center"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" media="(max-width: 480px)" srcset="/assets/img/soup_bad_example_12500-480.webp"></source> <source class="responsive-img-srcset" media="(max-width: 800px)" srcset="/assets/img/soup_bad_example_12500-800.webp"></source> <source class="responsive-img-srcset" media="(max-width: 1400px)" srcset="/assets/img/soup_bad_example_12500-1400.webp"></source> <img src="/assets/img/soup_bad_example_12500.png" class="img-fluid rounded z-depth-1" width="auto" height="auto" title="Soup missasignment example" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> An example of imbalance class in which 500 cells are from one Donor and 12000 from the remaining 6 donors. </div> <p>In another simulation of subsetting 500 cells from one donor, here is the result:</p> <div class="row" style="text-align: center"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" media="(max-width: 480px)" srcset="/assets/img/soup_bad_example2_12500-480.webp"></source> <source class="responsive-img-srcset" media="(max-width: 800px)" srcset="/assets/img/soup_bad_example2_12500-800.webp"></source> <source class="responsive-img-srcset" media="(max-width: 1400px)" srcset="/assets/img/soup_bad_example2_12500-1400.webp"></source> <img src="/assets/img/soup_bad_example2_12500.png" class="img-fluid rounded z-depth-1" width="auto" height="auto" title="Soup missasignment example" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Another example of imbalance class in which 500 cells are from one Donor and 12000 from the remaining 6 donors. </div> <p>While we have 500 cells fall nicely into one category, they were merged with cells from another donor :(. <strong>Imbalance scenario 2:</strong></p> <div class="row" style="text-align: center"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" media="(max-width: 480px)" srcset="/assets/img/soup_bad_example3_11500-480.webp"></source> <source class="responsive-img-srcset" media="(max-width: 800px)" srcset="/assets/img/soup_bad_example3_11500-800.webp"></source> <source class="responsive-img-srcset" media="(max-width: 1400px)" srcset="/assets/img/soup_bad_example3_11500-1400.webp"></source> <img src="/assets/img/soup_bad_example3_11500.png" class="img-fluid rounded z-depth-1" width="auto" height="auto" title="Soup missasignment example" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <div class="caption"> Another example of imbalance class in which 500 cells are from one Donor, 1000 from another and 10000 from the remaining 5 donors. </div> <p><strong>Using known genotypes to improve clustering</strong> Given that this problem is an optimization problem, where we are trying to find the center of the clusters, which are essentially the genotypes of the donors, we can try to feed the known genotype data to the algorithm, if such an information is available. Souporcell implements this by using the known genotype as the initial value for the sparse-mixture gaussian model. This makes a lot of sense! Moreover, hypothetically speaking, this should also solve the imbalance class situation as a good initial value has been shown to eleviate this. A good initial value might help us get closer to the true $\theta_i$.</p> <p>I attempted to create the genotype data for these 7 samples by using the paired-tissue data. While we can generate the genotype from the BAM files that we extract the cells from, this might cause some biases. With these genotypes information, here is the result in the 12500-cell class imbalance case:</p> <div class="row" style="text-align: center"> <div class="col-sm mt-3 mt-md-0"> <figure> <picture> <source class="responsive-img-srcset" media="(max-width: 480px)" srcset="/assets/img/soup_good_example_genotype2-480.webp"></source> <source class="responsive-img-srcset" media="(max-width: 800px)" srcset="/assets/img/soup_good_example_genotype2-800.webp"></source> <source class="responsive-img-srcset" media="(max-width: 1400px)" srcset="/assets/img/soup_good_example_genotype2-1400.webp"></source> <img src="/assets/img/soup_good_example_genotype2.jpg" class="img-fluid rounded z-depth-1" width="auto" height="auto" title="example image" onerror="this.onerror=null; $('.responsive-img-srcset').remove();"> </picture> </figure> </div> </div> <p><strong>What if some cells‚Äô identity are known ?</strong> In our lab, we used to generate the corresponding hashtag data of the cells from GEX experiments. The hashtag are not quite reliable to</p> <p>What do we learn from these examples?</p> <font size="5">1. **Conclusion** </font> <p>I think there are a few things I learned from these experiments with souporcell:</p> <ul> <li>the result given after a single run might be very inaccurate, and is driven by stochasticity. With that being said, rerunning an experiment with a different set of parameter might help</li> <li>If one can afford to genotype the samples, it is always better to include the known genotype as part of the run, instead of, for example, matching the found genotype taken from souporcell and match that with a known genotype later on</li> <li>I think we should also ask if imbalance class is a weakness of most unsupervised genetic demultiplexing tools out there.</li> </ul> </div> </article> <br> <hr> <br> <ul class="list-disc pl-8"></ul> <h2 class="text-3xl font-semibold mb-4 mt-12">Enjoy Reading This Article?</h2> <p class="mb-2">Here are some more articles you might like to read next:</p> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/WGCNA-vis/">üìà üåêVisualizing your WGCNA result</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/WGCNA-pip/">üì¢ Streamline your WGCNA analysis pipeline</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/sc-vis/">üìà Some visualization tips for scRNA-seq data</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/math/">Sparse-Mixture Gaussian clustering for genetic demultiplexing</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2021/attention/">üì¢ üìè Attention-based Transformer, a mathematical primer</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/PCA/">üì¢ Principle Component Analysis (PCA) - intuition</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/souporcell-GPU/">üì¢ Genetic-demultiplexing with Souporcell - GPU implementation</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/wPCA/">üì¢üìè Implementing weighted PCA in python</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/RV2/">üìè How do we measure correlation between matrices ?</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/NovaseqX/">üõ†Ô∏è üì¢ Generate FASTQ files for new NovaseqX sequencing data</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/projection/">üìè Projection in linear subspace</a> </li> </div> </div> <footer class="fixed-bottom"> <div class="container mt-0"> ¬© Copyright 2023 Hoang (Anh) Tran. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. Photos from <a href="https://unsplash.com" target="_blank" rel="external nofollow noopener">Unsplash</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@4/imagesloaded.pkgd.min.js"></script> <script defer src="/assets/js/masonry.js" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.0.8/dist/medium-zoom.min.js" integrity="sha256-7PhEpEWEW0XXQ0k6kQrPKwuoIomz8R8IYyuU1Qew4P8=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?7b30caa5023af4af8408a472dc4e1ebb"></script> <script defer src="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.js"></script> <script src="/assets/js/no_defer.js?d633890033921b33e0ceb13d22340a9c"></script> <script defer src="/assets/js/common.js?acdb9690d7641b2f8d40529018c71a01"></script> <script defer src="/assets/js/copy_code.js?07b8786bab9b4abe90d10e61f7d12ff7" type="text/javascript"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script type="text/javascript">window.MathJax={tex:{tags:"ams"}};</script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script> <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script async src="https://www.googletagmanager.com/gtag/js?id=G-1W7V2Y0MXW"></script> <script>function gtag(){window.dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-1W7V2Y0MXW");</script> <script async src="https://rum.cronitor.io/script.js"></script> <script>window.cronitor=window.cronitor||function(){(window.cronitor.q=window.cronitor.q||[]).push(arguments)},cronitor("config",{clientKey:""});</script> <script type="text/javascript">function progressBarSetup(){"max"in document.createElement("progress")?(initializeProgressElement(),$(document).on("scroll",function(){progressBar.attr({value:getCurrentScrollPosition()})}),$(window).on("resize",initializeProgressElement)):(resizeProgressBar(),$(document).on("scroll",resizeProgressBar),$(window).on("resize",resizeProgressBar))}function getCurrentScrollPosition(){return $(window).scrollTop()}function initializeProgressElement(){let e=$("#navbar").outerHeight(!0);$("body").css({"padding-top":e}),$("progress-container").css({"padding-top":e}),progressBar.css({top:e}),progressBar.attr({max:getDistanceToScroll(),value:getCurrentScrollPosition()})}function getDistanceToScroll(){return $(document).height()-$(window).height()}function resizeProgressBar(){progressBar.css({width:getWidthPercentage()+"%"})}function getWidthPercentage(){return getCurrentScrollPosition()/getDistanceToScroll()*100}const progressBar=$("#progress");window.onload=function(){setTimeout(progressBarSetup,50)};</script> </body> </html>