---
layout: post
title:  üìà üåêVisualizing your WGCNA result 
date: 2023-12-11 10:25:00
description: Learn to visualize your network module in an intuitive way
tags: visualization  
categories: Bioinformatics
related_posts: false
featured: true
thumbnail: /assets/img/network.png
---

**1. Volcano plot**

After the construction of gene network modules  with WGCNA, I guess one would be left with the question: what's next? As a no-brainer step, one can screen for interesting modules that correlate with clinical features of interest. For example, I have seen people using the following heatmap to summarize the result of correlating modules with multiple clinical features:

<div class="row" style="text-align: center">
    <div class="col-sm mt-3 mt-md-0">
        {% include figure.html path="assets/img/wgcna_heatmap.png" title="Lineage dotplot" class="img-fluid rounded z-depth-1" %}
    </div>
</div>
<div class="caption">
    A heatmap showing the correlation of different module's eigengenes and features of interest
</div>
However, for someone who would want to construct multiple modules from various cell-type, one can end up with a lot of network modules. In this case, a volcano plot might be useful to summarize the result of correlation between hundreds of modules and only a few variables (2-3 variables). For example, I identified quite a lot of module from just one population of epithelial cells obtained from scRNA-seq data alone:

<div class="row" style="text-align: center">
    <div class="col-sm mt-3 mt-md-0">
        {% include figure.html path="assets/img/wgcna_volcano.png" title="Lineage dotplot" class="img-fluid rounded z-depth-1" %}
    </div>
</div>
<div class="caption">
    A Volcano plot showing the correlation coefficient of different module's eigengenes (x-axis) and features of interest and their corresponding p-value (y-axis)
</div>

Of course, one can argue that the whole point of a volcano plot and a heatmap are very different. When you have a lot of clinical variables, a heatmap can be useful to highlight the bulk pattern of correlations between modules-variables as well as between variables. For only a few variable, a volcano plot might be more useful: we can do some more follow-up on the modules that pass a certain threshold for correlation. 

**2. Trend plot**

One way to "QC" a gene module as well as visualizing its correlating pattern with a response variable is to, welp, plot all the gene expressions and the variable on the same plot. This might sound like a lot, especially when you have hundreds if not thousands of genes. One can do the following: select genes that have  high module-memberships to the module, scale them to be between 0-1, and  plot only those genes' expressions (log10 CPM) against the values of the response variable; simultaneously, plotting the mean trend of the module as well, e.g. eigengene. For example, in our study, we detect an interesting module that is tightly correlated with immune-infiltration:

<div class="row" style="text-align: center">
    <div class="col-sm mt-3 mt-md-0">
        {% include figure.html path="assets/img/e5_example.png" title="Lineage dotplot" class="img-fluid rounded z-depth-1" %}
    </div>
</div>
<div class="caption">
    A scatterplot showing the correlation between a gene module in Epithelial cells and immune infiltration
</div>
You can also connect the dots with lines, so it will look more like a bunch of lines making up one overarching trend. 
It is quite interesting because the genes do look highly correlated to each other and to the mean-trend (the blue line). Simultaneously, the pattern is almost exponential. We actually detect these genes by correlating the log10 value of immune infiltration and the modules' eigengenes! 

From this experience, I also realize the importance of using Module's mean gene, as opposed to module's eigengene to represent the overall trend of a module, as suggested by the author. Obviously, these two values represent two different thing! the module's mean gene might be useful as we are built on the assumption that most genes are very similar and "wiggle" around a mean trend. The module's eigengene simply is the PC1 of the genes, and not necessarily their mean, but a weighted sum of the genes that has the most variation.

**3. Scatter plot**

Another kind of plot you can use alongside with the trend plot is the basic scatterplot on a per gene basis. For example, here is the log10CPM of genes highly correlating with log10(immune infiltration):
<div class="row" style="text-align: center">
    <div class="col-sm mt-3 mt-md-0">
        {% include figure.html path="assets/img/wgcna_scatterplot.png" title="Lineage dotplot" class="img-fluid rounded z-depth-1" %}
    </div>
</div>
<div class="caption">
    A scatterplot showing the correlation between a gene module in Epithelial cells and immune infiltration
</div>

It is important to report statistics such as the correlation coefficient, and one might also  add a SE band for each gene!


**4. Other thoughts**

I think it is important to use other kind of measures for follow-up analyses on interesting modules. One doesn't have to use module eigengenes to represent the overall trend of a module.  One can simply average the scaled values of all the genes (scaling is important here due to different dynamic range):

```R
moduleMeangenes=function(matrix,color) {
    if (length(color)!=ncol(matrix)) {
        print("the size of color vector doesnt match matrix size")
    }
     modules_genes=split(1:ncol(matrix),color)
     names(modules_genes)=paste0("ME",names(modules_genes))
        
     return(sapply(modules_genes,function(i) apply(scale(matrix[,i]),1,mean)))
    
}
```
here $$matrix$$ would be a sample $$\times$$ genes, and $$color$$ would be a vector assigning a group to each gene.


 At the same time, one can decide to use Pearson $$\rho$$, the Kendall correlation, or the Spearman correlation for screening interesting modules. Each is simply to answer a different kind of question. 
