---
layout: post
title: üõ†Ô∏è üì¢ Generate FASTQ files for new NovaseqX sequencing data
date: 2023-06-28
description: learn to trick your software with this tip
tags: formatting images
categories: bioinformatics
thumbnail: assets/img/illumina.png
---

If you happen to align the most recent Novaseq runs with the new technology, chances are that you will receive an error somewhere in cellranger-mkfastq

```bash
Log message:
Traceback (most recent call last):
File "/software/cellranger-6.0.1/external/martian/adapters/python/martian_shell.py", line 659, in _main
stage.main()
File "/software/cellranger-6.0.1/external/martian/adapters/python/martian_shell.py", line 618, in main
self._run(lambda: self._module.main(args, outs))
File "/software/cellranger-6.0.1/external/martian/adapters/python/martian_shell.py", line 589, in _run
cmd()
File "/software/cellranger-6.0.1/external/martian/adapters/python/martian_shell.py", line 618, in <lambda>
self._run(lambda: self._module.main(args, outs))
File "/software/cellranger-6.0.1/mro/tenkit/stages/make_fastqs/make_fastqs_preflight/__init__.py", line 53, in main
(rta_version, _, bcl_params) = tk_bcl.get_rta_version(args.run_path)
File "/software/cellranger-6.0.1/lib/python/tenkit/bcl.py", line 144, in get_rta_version
application_version = tree.getroot().find("ApplicationVersion").text
AttributeError: 'NoneType' object has no attribute 'text'
```

It seems that in the sequencing folder of this run doesnt have some attribute required by cellranger.
However, cellranger is just a wrapper of bcl2fastq, and we won't need the attribute 'text' to fix it anyway. So the solution, as suggested by the 10x website, is to manually generate the fastq from BCL files. 
I would copied the files to bcb-data and run bcl2fastq. The tricky part is that cellranger-mkfastq would do it dirty work to translate its samplesheet to bcl2fastq's samplesheet (these two samplesheets are different and you can't feed cellranger-mkfastq samplesheet to bcl2fastq). 

Here are two ways that you can generate the samplesheet for bcl2fastq:

1. Use an existing flowcell with the same chemistry and configuration (dual-index or single-index) as the one you obtained from Novaseq, and run cellranger on that flowcell. So for example, go to a directory that you want, and run:

```bash
cellranger mkfastq  \
-R /existing-flowcell-dir/230421_NB551582_0149_AHF3C2BGXT 
--sample-sheet /some-dir/samplesheet_for_new_run.csv --barcode-mismatches=0
```

samplesheet here would be your samplesheet for this NovaseqX run. 
You don't have to wait for this to run, just cancel it once bcl2fastq task is done. Inside your directory, you will have a folder of your flowcell, something like HF3C2BGXT. You can do 
```bash
 find . |grep csv
```
to find the bcl2fastq samplesheet: 
```bash
./MAKE_FASTQS_CS/MAKE_FASTQS/PREPARE_SAMPLESHEET/fork0/chnk0-u2a1f9c52b1/files/samplesheet.csv
./MAKE_FASTQS_CS/MAKE_FASTQS/PREPARE_SAMPLESHEET/fork0/chnk0-u2a1f9c52b1/files/input_samplesheet.csv
```
There are two files here, just use the one called samplesheet.csv and feed that into bcl2fastq:
```bash
bcl2fastq -i /projects/Hoang/230614_SL-EXC_0018_B223YNKLT3/Data/Intensities/BaseCalls/ -o /projects/Hoang/230614_SL-EXC_0018_B223YNKLT3/out/ 
--sample-sheet ./MAKE_FASTQS_CS/MAKE_FASTQS/PREPARE_SAMPLESHEET/fork0/chnk0-u2a1f9c52b1/files/samplesheet.csv 
```

You can do this, or you can manually  create the bcl2fastq samplesheet by search up 10x barcode files. This is tedious though... You have to search each index to get the write oligo barcodes. 
