<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title>ğŸ› ï¸ ğŸ“¢ Generate FASTQ files for new NovaseqX sequencing data | Hoang (Anh) Tran</title> <meta name="author" content="Hoang (Anh) Tran"> <meta name="description" content="learn to trick your software with this tip"> <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.css"> <link rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?19f3075a2d19613090fe9e16b564e1fe" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%E2%9A%9B%EF%B8%8F&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://hoangmgh.github.io/blog/2023/NovaseqX/"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?e74e74bf055e5729d44a7d031a5ca6a5" media="none" id="highlight_theme_dark"> <script src="/assets/js/theme.js?96d6b3e1c3604aca8b6134c7afdd5db6"></script> <script src="/assets/js/dark_mode.js?9b17307bb950ffa2e34be0227f53558f"></script> <script async src="https://www.googletagmanager.com/gtag/js?id={{%20site.google_analytics%20}}"></script> <script>function gtag(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-1W7V2Y0MXW");</script> </head> <body class="fixed-top-nav "> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"><span class="font-weight-bold">HoangÂ </span>(Anh)Â Tran</a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about</a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">Bioinformatics<span class="sr-only">(current)</span></a> </li> <li class="nav-item "> <a class="nav-link" href="/projects/">Arts</a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications</a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv</a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="fa-solid fa-moon"></i> <i class="fa-solid fa-sun"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5"> <div class="post"> <header class="post-header"> <h1 class="post-title">ğŸ› ï¸ ğŸ“¢ Generate FASTQ files for new NovaseqX sequencing data</h1> <p class="post-meta">June 28, 2023</p> <p class="post-tags"> <a href="/blog/2023"> <i class="fa-solid fa-calendar fa-sm"></i> 2023 </a> Â  Â· Â  <a href="/blog/tag/formatting"> <i class="fa-solid fa-hashtag fa-sm"></i> formatting</a> Â  <a href="/blog/tag/images"> <i class="fa-solid fa-hashtag fa-sm"></i> images</a> Â  Â  Â· Â  <a href="/blog/category/bioinformatics"> <i class="fa-solid fa-tag fa-sm"></i> bioinformatics</a> Â  </p> </header> <article class="post-content"> <div id="markdown-content"> <p>If you happen to align the most recent Novaseq runs with the new technology, chances are that you will receive an error somewhere in cellranger-mkfastq</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Log message:
Traceback <span class="o">(</span>most recent call last<span class="o">)</span>:
File <span class="s2">"/software/cellranger-6.0.1/external/martian/adapters/python/martian_shell.py"</span>, line 659, <span class="k">in </span>_main
stage.main<span class="o">()</span>
File <span class="s2">"/software/cellranger-6.0.1/external/martian/adapters/python/martian_shell.py"</span>, line 618, <span class="k">in </span>main
self._run<span class="o">(</span>lambda: self._module.main<span class="o">(</span>args, outs<span class="o">))</span>
File <span class="s2">"/software/cellranger-6.0.1/external/martian/adapters/python/martian_shell.py"</span>, line 589, <span class="k">in </span>_run
cmd<span class="o">()</span>
File <span class="s2">"/software/cellranger-6.0.1/external/martian/adapters/python/martian_shell.py"</span>, line 618, <span class="k">in</span> &lt;lambda&gt;
self._run<span class="o">(</span>lambda: self._module.main<span class="o">(</span>args, outs<span class="o">))</span>
File <span class="s2">"/software/cellranger-6.0.1/mro/tenkit/stages/make_fastqs/make_fastqs_preflight/__init__.py"</span>, line 53, <span class="k">in </span>main
<span class="o">(</span>rta_version, _, bcl_params<span class="o">)</span> <span class="o">=</span> tk_bcl.get_rta_version<span class="o">(</span>args.run_path<span class="o">)</span>
File <span class="s2">"/software/cellranger-6.0.1/lib/python/tenkit/bcl.py"</span>, line 144, <span class="k">in </span>get_rta_version
application_version <span class="o">=</span> tree.getroot<span class="o">()</span>.find<span class="o">(</span><span class="s2">"ApplicationVersion"</span><span class="o">)</span>.text
AttributeError: <span class="s1">'NoneType'</span> object has no attribute <span class="s1">'text'</span>
</code></pre></div></div> <p>It seems that in the sequencing folder of this run doesnt have some attribute required by cellranger. However, cellranger is just a wrapper of bcl2fastq, and we wonâ€™t need the attribute â€˜textâ€™ to fix it anyway. So the solution, as suggested by the 10x website, is to manually generate the fastq from BCL files. I would copied the files to bcb-data and run bcl2fastq. The tricky part is that cellranger-mkfastq would do it dirty work to translate its samplesheet to bcl2fastqâ€™s samplesheet (these two samplesheets are different and you canâ€™t feed cellranger-mkfastq samplesheet to bcl2fastq).</p> <p>Here are two ways that you can generate the samplesheet for bcl2fastq:</p> <ol> <li>Use an existing flowcell with the same chemistry and configuration (dual-index or single-index) as the one you obtained from Novaseq, and run cellranger on that flowcell. So for example, go to a directory that you want, and run:</li> </ol> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cellranger mkfastq  <span class="se">\</span>
<span class="nt">-R</span> /existing-flowcell-dir/230421_NB551582_0149_AHF3C2BGXT 
<span class="nt">--sample-sheet</span> /some-dir/samplesheet_for_new_run.csv <span class="nt">--barcode-mismatches</span><span class="o">=</span>0
</code></pre></div></div> <p>samplesheet here would be your samplesheet for this NovaseqX run. You donâ€™t have to wait for this to run, just cancel it once bcl2fastq task is done. Inside your directory, you will have a folder of your flowcell, something like HF3C2BGXT. You can do</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code> find <span class="nb">.</span> |grep csv
</code></pre></div></div> <p>to find the bcl2fastq samplesheet:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>./MAKE_FASTQS_CS/MAKE_FASTQS/PREPARE_SAMPLESHEET/fork0/chnk0-u2a1f9c52b1/files/samplesheet.csv
./MAKE_FASTQS_CS/MAKE_FASTQS/PREPARE_SAMPLESHEET/fork0/chnk0-u2a1f9c52b1/files/input_samplesheet.csv
</code></pre></div></div> <p>There are two files here, just use the one called samplesheet.csv and feed that into bcl2fastq:</p> <div class="language-bash highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bcl2fastq <span class="nt">-i</span> /projects/Hoang/230614_SL-EXC_0018_B223YNKLT3/Data/Intensities/BaseCalls/ <span class="nt">-o</span> /projects/Hoang/230614_SL-EXC_0018_B223YNKLT3/out/ 
<span class="nt">--sample-sheet</span> ./MAKE_FASTQS_CS/MAKE_FASTQS/PREPARE_SAMPLESHEET/fork0/chnk0-u2a1f9c52b1/files/samplesheet.csv 
</code></pre></div></div> <p>You can do this, or you can manually create the bcl2fastq samplesheet by search up 10x barcode files. This is tedious thoughâ€¦ You have to search each index to get the write oligo barcodes.</p> </div> </article> <br> <hr> <br> <ul class="list-disc pl-8"></ul> <h2 class="text-3xl font-semibold mb-4 mt-12">Enjoy Reading This Article?</h2> <p class="mb-2">Here are some more articles you might like to read next:</p> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/benchmarking-souporcell/">âš–ï¸ How class imbalance might affect genetic demultiplexing of scRNA-seq data</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/WGCNA-pip/">ğŸ“¢ Streamline your WGCNA analysis pipeline</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/WGCNA-vis/">ğŸ“ˆ ğŸŒVisualizing your WGCNA result</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/sc-vis/">ğŸ“ˆ Some visualization tips for scRNA-seq data</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/wPCA/">ğŸ“¢ğŸ“ Implementing weighted PCA in python</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/souporcell-GPU/">ğŸ“¢ Genetic-demultiplexing with Souporcell - GPU implementation</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/PCA/">ğŸ“¢ Principle Component Analysis (PCA) - intuition</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/math/">Sparse-Mixture Gaussian clustering for genetic demultiplexing</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2023/RV2/">ğŸ“ How do we measure correlation between matrices ?</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2022/projection/">ğŸ“ Projection in linear subspace</a> </li> <li class="my-2"> <a class="text-pink-700 underline font-semibold hover:text-pink-800" href="/blog/2021/attention/">ğŸ“¢ ğŸ“ Attention-based Transformer, a mathematical primer</a> </li> </div> </div> <footer class="fixed-bottom"> <div class="container mt-0"> Â© Copyright 2023 Hoang (Anh) Tran. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. Photos from <a href="https://unsplash.com" target="_blank" rel="external nofollow noopener">Unsplash</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@4/imagesloaded.pkgd.min.js"></script> <script defer src="/assets/js/masonry.js" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.0.8/dist/medium-zoom.min.js" integrity="sha256-7PhEpEWEW0XXQ0k6kQrPKwuoIomz8R8IYyuU1Qew4P8=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?7b30caa5023af4af8408a472dc4e1ebb"></script> <script defer src="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.js"></script> <script src="/assets/js/no_defer.js?d633890033921b33e0ceb13d22340a9c"></script> <script defer src="/assets/js/common.js?acdb9690d7641b2f8d40529018c71a01"></script> <script defer src="/assets/js/copy_code.js?07b8786bab9b4abe90d10e61f7d12ff7" type="text/javascript"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script type="text/javascript">window.MathJax={tex:{tags:"ams"}};</script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script> <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script async src="https://www.googletagmanager.com/gtag/js?id=G-1W7V2Y0MXW"></script> <script>function gtag(){window.dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-1W7V2Y0MXW");</script> <script async src="https://rum.cronitor.io/script.js"></script> <script>window.cronitor=window.cronitor||function(){(window.cronitor.q=window.cronitor.q||[]).push(arguments)},cronitor("config",{clientKey:""});</script> <script type="text/javascript">function progressBarSetup(){"max"in document.createElement("progress")?(initializeProgressElement(),$(document).on("scroll",function(){progressBar.attr({value:getCurrentScrollPosition()})}),$(window).on("resize",initializeProgressElement)):(resizeProgressBar(),$(document).on("scroll",resizeProgressBar),$(window).on("resize",resizeProgressBar))}function getCurrentScrollPosition(){return $(window).scrollTop()}function initializeProgressElement(){let e=$("#navbar").outerHeight(!0);$("body").css({"padding-top":e}),$("progress-container").css({"padding-top":e}),progressBar.css({top:e}),progressBar.attr({max:getDistanceToScroll(),value:getCurrentScrollPosition()})}function getDistanceToScroll(){return $(document).height()-$(window).height()}function resizeProgressBar(){progressBar.css({width:getWidthPercentage()+"%"})}function getWidthPercentage(){return getCurrentScrollPosition()/getDistanceToScroll()*100}const progressBar=$("#progress");window.onload=function(){setTimeout(progressBarSetup,50)};</script> </body> </html>