<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title>Sparse-Mixture Gaussian clustering for genetic demultiplexing | Hoang (Anh) Tran</title> <meta name="author" content="Hoang (Anh) Tran"> <meta name="description" content="An introduction to souporcell clustering with the mixture gaussian model"> <meta name="keywords" content="jekyll, jekyll-theme, academic-website, portfolio-website"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.css"> <link rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?19f3075a2d19613090fe9e16b564e1fe" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%E2%9A%9B%EF%B8%8F&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://hoangmgh.github.io/blog/2023/math/"> <link rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?e74e74bf055e5729d44a7d031a5ca6a5" media="none" id="highlight_theme_dark"> <script src="/assets/js/theme.js?96d6b3e1c3604aca8b6134c7afdd5db6"></script> <script src="/assets/js/dark_mode.js?9b17307bb950ffa2e34be0227f53558f"></script> </head> <body> <p> {% if jekyll.environment == 'production' and site.google_analytics %} {% include analytics.html %} {% endif %} </p> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"><span class="font-weight-bold">Hoang </span>(Anh) Tran</a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about</a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">Bioinformatics<span class="sr-only">(current)</span></a> </li> <li class="nav-item "> <a class="nav-link" href="/projects/">Arts</a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications</a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv</a> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="fa-solid fa-moon"></i> <i class="fa-solid fa-sun"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5"> <div class="post"> <header class="post-header"> <h1 class="post-title">Sparse-Mixture Gaussian clustering for genetic demultiplexing</h1> <p class="post-meta">October 16, 2023</p> <p class="post-tags"> <a href="/blog/2023"> <i class="fa-solid fa-calendar fa-sm"></i> 2023 </a>   ·   <a href="/blog/tag/math"> <i class="fa-solid fa-hashtag fa-sm"></i> math</a>   <a href="/blog/tag/genetic-demultiplexing"> <i class="fa-solid fa-hashtag fa-sm"></i> genetic-demultiplexing</a>     ·   <a href="/blog/category/bioinformatics"> <i class="fa-solid fa-tag fa-sm"></i> Bioinformatics</a>   </p> </header> <article class="post-content"> <div id="markdown-content"> <p>I attempted to analyze the source code of <a href="https://github.com/wheaton5/souporcell" rel="external nofollow noopener" target="_blank">souporcell</a> and finally was able to understand its singlet identification step <a href="https://github.com/wheaton5/souporcell/blob/master/souporcell_pipeline.py" rel="external nofollow noopener" target="_blank">souporcell.py</a>. From my experience, the source code is much more straightforward to re-engineer than the mathematical formula given in the published paper, which might be a bit confusing.</p> <p>At the very core, souporcell attempts to model the genotype of each cell as a vector \(v_c\) of size \(1 \times l\), whose entry is calculated as \begin{equation} v_{cl} =\frac{r_{cl}}{a_{cl}+r_{cl}} <br> \end{equation}</p> <p>That is, we represented a cell’s genotype by a vector, whose entries contains the fraction of reads that support the reference loci. If we have \(8\) reads mapped to a loci \(l\), where \(3\) reads show evidence for the reference genotype and \(5\) for the alternative, this value is \(3/8\), for example. The entries of this vector can take in the value of \(0\), \(0.5\), or \(1\), corresponding to the \(0/0\), \(0/1\) and \(1/1\) genotype. However, due to noise and coverage, the value can be skewed away.</p> <p><strong>1. Mixture Gaussian model</strong></p> <p>Assuming that each cell has its genotype data generated by vartrix, we can generate the vector for each cell by using the <strong>ref.mtx</strong> and the <strong>alt.mtx</strong> files given by souporcell. Now, the problem of assigning a cell an identity is equated to clustering the given cells into \(K\) group. Since single-cell data is too sparse, any methods that measure distance between cells might not be very effective. Moreover, we can think of the set of vectors \(v_c\) will cluster into \(K\) group, corresponding to \(K\) donors, and the center of these \(K\) groups will be the donor’s genotypes. We can use the Mixture-Gaussian method to solve this problem.</p> <p>Assuming that each cell (observation) is independently drawn from a mixture of \(K\) components, then the probability of seeing the vector \(v_c\) is:</p> <p>\begin{equation} P(v_c)= \sum_{i=1}^8 \pi_i N( \theta_i| K=i) = \sum_{i=1}^8 P(K=i) N(v=v_c |\theta_i , K=i) \end{equation}</p> <p>Here we can think of this as a two-fold process: for each component \(k\), we pick it with the probability \(P(K=i)\). Each component is “tagged” with a probability distribution. In our case, the distribution is Multivariate-Gaussian, so it is given by \(N(v=v_c ,\theta_i)\). A Multivariate-Gaussian take in two parameters: the center vector \(\theta_i\) and a covariance matrix \(\Sigma\). For simplicity, I think the author assume that most genes are pair-wise equivariant. The probability of “picking” the components must sum up to \(1\), and hence \(P(K=i)\) represents the weight of the component. Here, the author also assumes that the probability of drawing from each component is equal, so we can drop the terms \(P(K=i)\), and our probability \(P(v_c)\) can be thought of as an averaging of \(K\) probabilities.</p> <p>The formula for \(N(v=v_c \| \theta_i)\) is also well-defined:</p> <p>\begin{equation} N(v=v_c | \theta_i ,K=i) = e^{-\frac{1}{2}|v_c-\theta_i|_2 } \end{equation}</p> <p>The further your vector \(v_c\) from the centre \(\theta_i\), the smaller the \(l_2\) distance, and hence the bigger the probability.s When the distance is large enough, the exponent approach \(-\infty\), and hence the probability approaches \(e^{-\infty} \longrightarrow 0\)</p> <p><strong>2. Sparse implementation</strong></p> <p>Now, it is not very straightforward to calculate \(\|v_c-\theta_i\|\), given that not all loci of cell are observed. That is, most of entries of \(v_c\) wont have values, due to the fact that we don’t see any reads mapped to the loci (\(r_{cl}=a_{cl}=0\)) at loci \(l\). To deal with this situation, the author proposed the “sparse” mixture implementation.</p> <p>Now, we can substitute the distance \(\|v_c-\theta_i\|\) with a version where only observed values contribute to the distance. To find the set of \(\theta_1\), \(\theta_2\),…\(\theta_K\), we use the Gradient-descent method to optimize our loss function.</p> <p>Note that the problem is now in fact an <strong>optimization</strong> problem, where a good set of parameters \(\theta_i\) is the one that maximize the probability of seeing the independent observations. Hence, we will be finding the centre vectors such that the following is maximized: \begin{equation} l(\theta_1,\theta_2,…,\theta_K)= \Pi_{i=1}^c P(v_i | \theta_1,\theta_2,..,\theta_K) \end{equation} which is also equivalent to maximizing the log (hence we call it the log-loss function) \begin{equation} log l(\theta)= log P(v_1 )+…+ log P(v_2) \end{equation}</p> </div> </article> </div> </div> <footer class="fixed-bottom"> <div class="container mt-0"> © Copyright 2023 Hoang (Anh) Tran. Powered by <a href="https://jekyllrb.com/" target="_blank" rel="external nofollow noopener">Jekyll</a> with <a href="https://github.com/alshedivat/al-folio" rel="external nofollow noopener" target="_blank">al-folio</a> theme. Hosted by <a href="https://pages.github.com/" target="_blank" rel="external nofollow noopener">GitHub Pages</a>. Photos from <a href="https://unsplash.com" target="_blank" rel="external nofollow noopener">Unsplash</a>. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@4/imagesloaded.pkgd.min.js"></script> <script defer src="/assets/js/masonry.js" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.0.8/dist/medium-zoom.min.js" integrity="sha256-7PhEpEWEW0XXQ0k6kQrPKwuoIomz8R8IYyuU1Qew4P8=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?7b30caa5023af4af8408a472dc4e1ebb"></script> <script defer src="https://unpkg.com/bootstrap-table@1.22.1/dist/bootstrap-table.min.js"></script> <script src="/assets/js/no_defer.js?d633890033921b33e0ceb13d22340a9c"></script> <script defer src="/assets/js/common.js?acdb9690d7641b2f8d40529018c71a01"></script> <script defer src="/assets/js/copy_code.js?07b8786bab9b4abe90d10e61f7d12ff7" type="text/javascript"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script type="text/javascript">window.MathJax={tex:{tags:"ams"}};</script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.0/es5/tex-mml-chtml.js"></script> <script defer src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script> <script async src="https://www.googletagmanager.com/gtag/js?id=G-1W7V2Y0MXW"></script> <script>function gtag(){window.dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],gtag("js",new Date),gtag("config","G-1W7V2Y0MXW");</script> <script async src="https://rum.cronitor.io/script.js"></script> <script>window.cronitor=window.cronitor||function(){(window.cronitor.q=window.cronitor.q||[]).push(arguments)},cronitor("config",{clientKey:""});</script> <script type="text/javascript">function progressBarSetup(){"max"in document.createElement("progress")?(initializeProgressElement(),$(document).on("scroll",function(){progressBar.attr({value:getCurrentScrollPosition()})}),$(window).on("resize",initializeProgressElement)):(resizeProgressBar(),$(document).on("scroll",resizeProgressBar),$(window).on("resize",resizeProgressBar))}function getCurrentScrollPosition(){return $(window).scrollTop()}function initializeProgressElement(){let e=$("#navbar").outerHeight(!0);$("body").css({"padding-top":e}),$("progress-container").css({"padding-top":e}),progressBar.css({top:e}),progressBar.attr({max:getDistanceToScroll(),value:getCurrentScrollPosition()})}function getDistanceToScroll(){return $(document).height()-$(window).height()}function resizeProgressBar(){progressBar.css({width:getWidthPercentage()+"%"})}function getWidthPercentage(){return getCurrentScrollPosition()/getDistanceToScroll()*100}const progressBar=$("#progress");window.onload=function(){setTimeout(progressBarSetup,50)};</script> </body> </html>